synapse_antispam/mjolnir/matching.py

```diff
@@ -0,0 +1,73 @@
+# -*- coding: utf-8 -*-
+# Copyright 2022 The Matrix.org Foundation C.I.C.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Tools in this file were copied from Synapse as these functions will not
+# remain publicly accessible in the module, see https://github.com/matrix-org/mjolnir/pull/174
+
+import re
+from typing import Pattern
+
+def re_word_boundary(r: str) -> str:
+    """
+    Adds word boundary characters to the start and end of an
+    expression to require that the match occur as a whole word,
+    but do so respecting the fact that strings starting or ending
+    with non-word characters will change word boundaries.
+    """
+    # we can't use \b as it chokes on unicode. however \W seems to be okay
+    # as shorthand for [^0-9A-Za-z_].
+    return r"(^|\W)%s(\W|$)" % (r,)
```

Comment:
Note that the match isn't exactly the same with and without `\b`:

![image](https://user-images.githubusercontent.com/5547783/153197608-b1cab40f-1419-4568-8209-dda58f0cc865.png)

![image](https://user-images.githubusercontent.com/5547783/153197629-13ef35ae-8005-4bd6-8292-10c46277772f.png)

But not sure if that's relevant enough to care about it.

Reply:
At this stage, this is purely copy&paste from Synapse. I think we can refine later if needs be.